src/momentum/PSAR.ts

```diff
@@ -0,0 +1,128 @@
+
+  If long:
+  Find the difference between the highest price made while in the trade and the SAR for today.
+  Multiple the difference by the AF and ADD the result to today's SAR to obtain the SAR for tomorrow.
+  Use 0.02 for the first AF and increase it by 0.02 on every day that a new high for the trade is made.
+  If a new high is not made continue to use the AF as last increased. Do not increase the AF above .20
+
+  Never move the SAR for tomorrow ABOVE the previous day's LOW or today's LOW.
+  If the SAR is calculated to be ABOVE the previous day's LOW or today's LOW then use the lower low between today and the previous day as the new SAR.
+  Make the next day's calculations based on this SAR.
+
+  If short:
+  Find the difference between the lowest price made while in the trade and the SAR for today.
+  Multiple the difference by the AF and SUBTRACT the result to today's SAR to obtain the SAR for tomorrow.
+  Use 0.02 for the first AF and increase it by 0.02 on every day that a new high for the trade is made.
+  If a new high is not made continue to use the AF as last increased. Do not increase the AF above .20
+
+  Never move the SAR for tomorrow BELOW the previous day's HIGH or today's HIGH.
+  If the SAR is calculated to be BELOW the previous day's HIGH or today's HIGH then use the higher high between today and the previous day as the new SAR. Make the next day's calculations based on this SAR.
+  ----------------------------------------------------------------------------------------
+*/
+export class PSARInput extends IndicatorInput{
+  step:number;
+  max:number;
+  high:number[];
+  low:number[];
+};
+
+export class PSAR extends Indicator {
+  result:number[];
+  generator:IterableIterator<number | undefined>;
+  constructor (input:PSARInput) {
+    super(input);
+
+    let highs = input.high || [];
+    let lows = input.low || [];
+
+    var genFn = function* (step:number, max:number):IterableIterator<number | undefined> {
+      let curr, extreme, sar, furthest;
+
+      let up = true;
+      let accel = step;
+      let prev = yield;
+      while(true) {
+        if (curr) {
+          sar = sar + accel * (extreme - sar);
+
+          if (up) {
+            sar = Math.min(sar, furthest.low, prev.low);
+
+            if (curr.high > extreme) {
+              extreme = curr.high;
+              accel = Math.min(accel + step, max);
+            };
+          } else {
+            sar = Math.max(sar, furthest.high, prev.high);
+
+            if (curr.low < extreme) {
+              extreme = curr.low;
+              accel = Math.min(accel + step, max);
+            }
+          }
+
+          if ((up && curr.low < sar) || (!up && curr.high > sar)) {
+            accel = step;
+            sar = extreme;
+            up = !up;
+
+            extreme = !up ? curr.low : curr.high;
+          }
+        } else {
+          // Randomly setup start values? What is the trend on first tick??
+          sar = prev.low; extreme = prev.high;
```

Comment:
On first entry not too sure what should be set. In the tulip indicators it checks against the first and second entry. Just wasn't convinced about if it mattered and the logic was uglier so did this atm...